FEATURE: Custom grouping patterns

OsamaSayegh opened the pull request:
This PR adds support for message grouping based on regular expression patterns that can be specified via the UI at the `/settings` page. Existing and new messages that match a pattern will be grouped together and displayed as a single row which will have the same timestamp as the most recent message in the group.

Removing a pattern will undo the grouping and the individual messages in the group will go back to having their own rows in the UI.

Screenshot:

![image](https://user-images.githubusercontent.com/17474474/70458652-95a84800-1ac3-11ea-949d-e0d809759361.png)

cc @SamSaffron

lis2 commented on client-app/app/components/patterns-list.js:
```diff
@@ -1,11 +1,13 @@
   immutable: not("mutable"),
+  showCounter: equal("key", "suppression"),
+  showResetCounter: equal("key", "suppression"),
 
```
I don't follow that, if `showCounter` and `showResetCounter` got same value, 
do we need 2 conditional statements?

```
    {{#if showCounter}}
      <input disabled=true title="Number of logs that have been suppressed by this pattern" class="count shrink" value={{pattern.count}}>
    {{/if}}
    {{#if showResetCounter}}
      <button {{action "resetCount" pattern}} class="btn no-text reset shrink" title="Reset count of suppressed logs" disabled={{pattern.zeroCount}}>
        <i class="fa fa-rotate-right"></i>
      </button>
    {{/if}}
```

lis2 commented on lib/logster/base_store.rb:
```diff
@@ -131,6 +131,18 @@ def rate_limited?(ip_address, perform: false, limit: 60)
+    # removes the Logster::Group instance associated with the given pattern
+    def remove_pattern_group(pattern)
+    end
+
```
Do we want to force children classes to implement those methods by throwing `not_implemented`?

lis2 commented on lib/logster/group.rb:
```diff
@@ -0,0 +1,126 @@
+    def initialize(key, messages_keys = [], timestamp: nil)
+      @key = key
+      @messages_keys = messages_keys || []
+      @timestamp = timestamp || 0
```
I got small cosmetic suggestion here. <aybe it would be better to pass 0 as default to timestamp so we could be able to skip `|| 0`

lis2 commented on lib/logster/group.rb:
```diff
@@ -0,0 +1,126 @@
+
+    def self.max_size=(num) # used in tests
+      @max_size = num
+    end
```
I don't have a very strong opinion here, but I would like to avoid setting "production" code to have the nice syntax in the test.
Maybe, we can remove that method and in test use uglier https://apidock.com/ruby/Module/class_variable_set

OsamaSayegh replied:
I've made this change but I went with instance variables for the class rather than class variables because class variables are shared with instances of the class and I don't think instances need to have access to it.

lis2 commented on lib/logster/pattern.rb:
```diff
@@ -1,7 +1,15 @@
     class PatternError < StandardError; end
 
+    def self.inherited(subclass)
+      Logster::Pattern::ALL << subclass
```
ALL can be a little bit misleading because this is a constant, but dynamic :) What do you think of using a class variable?)

jjaffeux commented on client-app/app/components/env-tab.js:
```diff
@@ -4,27 +4,27 @@ import { buildHashString } from "client-app/lib/utilities";
-        {},
-        this.get("message.env")[this.get("current") - 1]
-      );
+      const currentEnv = Em.$.extend({}, this.get("currentEnv"));
```
do we need the `.get` here ?

jjaffeux commented on client-app/app/components/page-nav.js:
```diff
@@ -0,0 +1,45 @@
+import { later } from "@ember/runloop";
+
+export default Component.extend({
+  tagName: "div",
```
div  is the default

jjaffeux commented on client-app/app/components/page-nav.js:
```diff
@@ -0,0 +1,45 @@
+
+  didReceiveAttrs() {
+    this._super(...arguments);
+    later(this, () => {
```
Multiple notes here:
- layer without a delay is a smell to me, it reads like "couldn't make it work so I added a small delay and it works". :p wouldn't and afterRender be better?
- you should cancel any later in a willDestroyElement, otherwise this might create error when the component gets destroyed and later hasn't occured yet
- this could be inline: `later(this, () => this.navigate(this.position))`

OsamaSayegh replied:
A big chunk of the code in this PR is about 1 year old, including this override to the `didReceiveAttrs` method. I've removed this override entirely since I've refactored this component and the override is no longer needed 👍

jjaffeux commented on client-app/app/components/page-nav.js:
```diff
@@ -0,0 +1,45 @@
+    this.navigate(0);
+  }),
+
+  disableBackButtons: computed("position", function() {
```
what about using:

```
equal("position", 0)
```

jjaffeux commented on client-app/app/models/group.js:
```diff
@@ -0,0 +1,23 @@
+    this.set("messages", messages);
+  },
+
+  displayMessage: computed("regex", function() {
```
`displayMessage: reads("regex")`

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -1,40 +1,112 @@
+  },
+
+  tabChanged(newTab) {
+    this.set("currentTab", newTab);
```
group with setProperties

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -1,40 +1,112 @@
+      row.key === this.currentRow.key
+    ) {
+      messageIndex = row.messages
+        .map(m => m.key)
```
mapBy("key")

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -1,40 +1,112 @@
+
+  updateSelectedRow() {
+    const currentKey = this.get("currentRow.key");
+    const rows = this.rows;
```
This one feels a little bit unecessary

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -43,13 +115,13 @@ export default Em.Object.extend({
     };
 
-    const search = this.get("search");
+    const search = this.search;
```
this one feels unnecessary

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -126,43 +198,47 @@ export default Em.Object.extend({
   hideCountInLoadMore: computed("search", "filter", function() {
-    const search = this.get("search");
-    const filter = this.get("filter");
+    const search = this.search;
```
this is minor nitpick but I don't feel like we should have an intermediate const for this kind of cases, totally fine to have this.search in code

jjaffeux commented on client-app/app/models/message-collection.js:
```diff
@@ -126,43 +198,47 @@ export default Em.Object.extend({
+    const rows = this.rows;
+    const firstLog = rows[0];
+    const firstKey = firstLog.group ? firstLog.row_id : firstLog.key;
+    const includedGroups = rows.filterBy("group").map(g => g.regex);
```
mapBy("regex")

jjaffeux commented on client-app/app/models/group.js:
```diff
@@ -0,0 +1,23 @@
+export default EmberObject.extend({
+  selected: false,
+  showCount: true,
+  key: alias("regex"),
```
if we don't need to mutate key, reads is probably a better choice

jjaffeux commented on client-app/app/components/page-nav.js:
```diff
@@ -0,0 +1,45 @@
+    });
+  },
+
+  resetPosition: observer("resetPositionOnChange", function() {
```
Can't we get read of this observer?

OsamaSayegh replied:
Yes, I've removed it. When I picked up this PR again I refactored it to make it follow the DDAU practice as much as possible, and that made removing this observer super easy since you know only one place can modify an attribute and you know what that place is.

jjaffeux commented on client-app/app/controllers/index.js:
```diff
@@ -132,19 +123,17 @@ export default Controller.extend({
 
   filterChanged: observer("filter.length", function() {
```
This is not from this PR but we can certainly get rid of this observer

SamSaffron commented:
@lis2 / @jjaffeux  is this good to merge now? Can't wait to try this out on dev! Such an awesome feature.

OsamaSayegh commented:
Thanks! ❤️ I've addressed the comments here and the feedback on dev (show the most recent message rather than the regex) and I'm happy where this is at now. Will merge tomorrow if I don't hear anything from @lis2 or @jjaffeux until then. Thank you guys for the review! ❤️
